### PR TITLE
[EDL][FIX] Fix EDL cut actions

### PR DESCRIPTION
**Pull request by enen92**
## Description
Currently EDL cut actions are broken - when reaching EDL cut start times, Kodi seeks to incorrect positions on the timeline (e.g. in the provided example Kodi seeks to 10 minutes playback, shows 10 minutes on the seek bar but the position on the actual file is far greater than 10 min).

In simple terms, EDL cuts remove playback time from the actual file. So, for instance, if you have a media item with 1 hour duration and a 30 minute EDL cut, the displayed total time should be 30 minutes (as if the cut was not part of the original item) and Kodi should be able to remove and restore the playback position as you play the file. In the provided example, at 10 seconds kodi should jump to 10 minutes of the original file but still display 11 seconds of playback on the next second.
This is what is supposed to happen with this PR. Plus a few improvements (todos and fixmes)/code cleanup were also applied.

There are a couple of issues (not introduced with this PR) that would be nice to get fixed in the near future:

- Edl cuts are displayed in the `Player.Cutlist` (ranges control in estuary), which is wrong. The ranges should not include cuts as the video timeline is no longer the original one. I will address this in the future by deprecating `Player.CutList` infolabel and introducing new infolabels. Hopefully `Player.CutList` can be safely removed in v21.

~- If you fast forward the playback with really high fast forward ratios (e.g. 8x or 16x) and then resume playback by hitting play - video will freeze. This is an extremely edge use case that I doubt anyone will use. Regardless, it's possible and it would be nice to get it working. Unfortunately, I feel @FernetMenta is the only one which can help here. The same happens when fast-forwarding backwards. Your help, as always is greatly appreciated. There's something missing after resuming from high fast forward ratios that I'm not able to figure out.~ (Fixed in https://github.com/xbmc/xbmc/pull/20744)

@a1rwulf since you were the one reporting the original "issue", it would be nice if you could spare some time testing and functionally validating this.

## Motivation and context
Fix EDL cuts

## How has this been tested?
Mainly with the following EDL file which contains two cuts and a commbreak:

```
10    600     0
610   720     0
730   749     3
```

Kodi should skip (590 seconds) at 10s and continue playback (11 secs playback) as if the first cut was not part of the original file. Should skip again at 20s playback. Total play duration should be equal to the file duration minus the time dedicated to EDL cuts.

## What is the effect on users?
EDL cuts should work again and skip to the appropriate future positions on the playback timeline.

### CODE REVIEW

**Review by lrusak — COMMENTED** [review]
Can't say much about the functionality of the code so I just left a couple of comments about what I saw.

It really looks like `CEdl` would benefit a lot from using `std::chrono`. This would remove a lot of the ambiguity when dealing with time.

File: xbmc/cores/VideoPlayer/Edl.cpp [hunk 1]
```diff
@@ -723,55 +723,57 @@ bool CEdl::HasEdits() const
...
-  // since if the file HasEdits but not any EDL cut there's
-  // no time to restore
-  if (!HasEdits())
-    return dClock;
+  if (!HasCuts())
+    return seek;
 
-  double dSeek = dClock;
-  for (size_t i = 0; i < m_vecEdits.size(); ++i)
+  for (const EDL::Edit edit : m_vecEdits)
```
[the remark this answers is not in the document]
Comment (enen92) [reply xbmc/cores/VideoPlayer/Edl.cpp]: agrr, good eye! Not sure how I missed this one :)
Thanks for the suggestions

File: xbmc/cores/VideoPlayer/VideoPlayer.cpp [hunk 2]
```diff
@@ -4738,13 +4758,10 @@ void CVideoPlayer::UpdatePlayState(double timeout)
...
-  // FIXME - This block of code only makes sense if the item has EDL *cuts*
-  // not any of the other edit types (mute, commbreak, etc). We loop over
-  // the EDL edit list twice unnecessarily.
-  if (m_Edl.HasEdits())
+  if (m_Edl.HasCuts())
   {
-    state.time = (double) m_Edl.RemoveCutTime(llrint(state.time));
-    state.timeMax = (double) m_Edl.RemoveCutTime(llrint(state.timeMax));
+    state.time = static_cast<double>(m_Edl.GetTimeWithoutCuts(std::lround(state.time)));
+    state.timeMax = std::lround(state.timeMax - static_cast<double>(m_Edl.GetTotalCutTime()));
```
Comment (lrusak) [inline xbmc/cores/VideoPlayer/VideoPlayer.cpp:4764]: This seems wrong, `timeMax` is a `double`, did you mean to use `std::round` instead (or nothing at all perhaps)?

Comment (enen92) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:4764]: Round to the nearest integer according to the original implementation (llrint). lround was your recommendation on the last review

Comment (lrusak) [reply xbmc/cores/VideoPlayer/VideoPlayer.cpp:4764]: Yea sure, it works for the previous line but on this line do you even need to round at all?

**Comment by FernetMenta** [comment]
@enen92 do you have a log showing the freeze?

**Comment by enen92** [comment]
@FernetMenta this seems to happen even without EDL files.

In the following log I fast forward at 32x speed and then resume by hitting play:

http://paste.kodi.tv/uwinuretod

Fast forwarding works well, but after resume Kodi sort of freezes (plays like 3 frames per second, no audio).

The exact same happens if you "fast forward backwards" and hit resume:

http://paste.kodi.tv/jigafuyava

While for backwards it "breaks" right away, fast forwarding at small steps (2x, 4x) and then resume works well.

This video illustrates the issue well: https://www.youtube.com/watch?v=aJuav4CiFcs

**Comment by FernetMenta** [comment]
@enen92 some very old code + same fixes for edge cases lead to this issue. Dropping packets in demuxer stage does not make much sense: https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDDemuxers/DVDDemuxFFmpeg.cpp#L741
Only very few containers/inputstreams want to be notified about speed from player.  Setting speed on demuxer had been dropped already, then resurrected. Kodi uses buffering/caching, hence reading speed from demuxer does never equal playback speed exactly.

I don't know why demuxer speed is reset in flush. As a result demuxer speed was not reset when going back to normal speed. 
```
@@ -3854,6 +3856,8 @@ void CVideoPlayer::FlushBuffers(double pts, bool accurate, bool sync)
   UpdatePlayState(0);

   m_demuxerSpeed = DVD_PLAYSPEED_NORMAL;
+  if (m_pDemuxer)
+    m_pDemuxer->SetSpeed(DVD_PLAYSPEED_NORMAL);

 }
```

**Comment by enen92** [comment]
Sorry for the small delay @FernetMenta. Life decided to be everything except easy on the last days...

I can confirm your patch fixes the issue (with and without EDL cuts). Submitted your patch on another PR (https://github.com/xbmc/xbmc/pull/20744) as I realized it was not related to the work being done in this one. Thank you very much for the help.
Wishes of a happy new year for you and your relatives.

**Review by ksooo — APPROVED** [review]
LGTM

**Pull request merged**